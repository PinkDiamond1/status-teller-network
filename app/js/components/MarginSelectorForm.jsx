import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {FormGroup, Label, InputGroup, InputGroupAddon, Alert} from 'reactstrap';
import Input from 'react-validation/build/input';
import {withNamespaces} from 'react-i18next';
import {withRouter} from 'react-router-dom';
import Form from 'react-validation/build/form';
import {isInteger, required} from '../validators';

class MarginSelectorForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: '',
      marketError: '',
      isAbove: props.margin.isAbove,
      rate: props.margin.rate
    };

    // Redirecting to previous step
    if(!props.fiat) this.props.history.push('/seller/fiat');
  }

  onRadioButtonChange = (e) => {
    this.setState({isAbove: e.target.value === "above"});
  }

  onInputChange = (e) => {
    this.setState({rate: e.target.value});
  }

  _submit(address){
    if(!this.validate()) return;    
    this.props.onSubmit(parseInt(this.state.rate, 10), this.state.isAbove); 
    this.props.history.push(address);
  }

  back = () => {
    this.props.history.push("/seller/fiat");
  }

  continue = () => {
    // TODO: redirect to next step 
    this._submit("/seller/margin");
  };

  validate() {
    this.setState({error: '', marketError: ''});

    this.form.validateAll();
    if (this.form.getChildContext()._errors.length > 0) {
      this.setState({error: this.props.t('marginSelectorForm.error')});
      return false;
    }

    this.setState({error: ''});
    return true;
  }

  render() {
    const {t, fiat} = this.props;
    const {error, rate, isAbove} = this.state;

    if(!fiat) return null;

    return ( 
      <Form ref={c => { this.form = c; }}>
        {error && <Alert color="danger">{error}</Alert>}
        <FormGroup>
            <Label for="rate">{t('marginSelectorForm.margin')}</Label>
            <InputGroup>
              <Input type="text" name="rate" id="rate" placeholder="0" className="form-control"
                  value={rate} onChange={this.onInputChange} validations={[required, isInteger]} />
              <InputGroupAddon addonType="append">%</InputGroupAddon>
            </InputGroup>
        </FormGroup>
        <FormGroup>
          <Label check>
            <Input type="radio" name="market" value="above" onChange={this.onRadioButtonChange} checked={isAbove} /> {t('marginSelectorForm.aboveMarket')}
          </Label>
          <Label check>
            <Input type="radio" name="market" value="below" onChange={this.onRadioButtonChange} checked={!isAbove} /> {t('marginSelectorForm.belowMarket')}
          </Label>
        </FormGroup>

        <h3>{t('marginSelectorForm.sellPrice')}</h3>
        <div>
          1 TODO = 1,234.00 {fiat}
        </div>
        <small>{t('marginSelectorForm.priceOrigin')}</small>


        <h3>{t('marginSelectorForm.ourFee')}</h3>
        <div>
          TODO SNT
        </div>

        <a onClick={this.back}>{t("marginSelectorForm.back")}</a>
        <br />
        <a onClick={this.continue}>{t("marginSelectorForm.continue")}</a>
      </Form>
    );
  }
}

MarginSelectorForm.propTypes = {
  t: PropTypes.func,
  error: PropTypes.string,
  history: PropTypes.object,
  margin: PropTypes.object,
  onSubmit: PropTypes.func,
  fiat: PropTypes.string
};

export default withRouter(withNamespaces()(MarginSelectorForm));
