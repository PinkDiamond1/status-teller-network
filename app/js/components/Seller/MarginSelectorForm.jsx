import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {FormGroup, Label, InputGroup, InputGroupAddon, ButtonGroup, InputGroupText} from 'reactstrap';
import Input from 'react-validation/build/input';
import {withNamespaces} from 'react-i18next';
import Form from 'react-validation/build/form';
import {isNumber, required, lowerThan} from '../../validators';
import MarketButton from '../ui/MarketButton';

const ABOVE = 0;
const BELOW = 1;

class MarginSelectorForm extends Component {

  render() {
    const {t, currency, margin, marketType} = this.props;

    return (
      <Form ref={c => { this.form = c; }}>
        <h2>{t('sellerMarginContainer.title')}</h2>
        <FormGroup>
            <Label for="margin">{t('marginSelectorForm.margin')}</Label>
            <InputGroup className="full-width-input">
              <Input type="number"
                     name="margin"
                     id="margin"
                     placeholder="0"
                     className="form-control prepend"
                     value={margin}
                     onChange={(e) => this.props.marginChange(e.target.value)}
                     validations={[required, isNumber, lowerThan.bind(null, 100)]} />
              <InputGroupAddon addonType="append"><InputGroupText>%</InputGroupText></InputGroupAddon>
            </InputGroup>
        </FormGroup>
        <FormGroup>
          <ButtonGroup className="d-flex">
            <MarketButton onClick={() => this.props.marketTypeChange(ABOVE)} active={marketType === ABOVE}>
              {t('marginSelectorForm.aboveMarket')}
            </MarketButton>
            <MarketButton onClick={() => this.props.marketTypeChange(BELOW)} active={marketType === BELOW}>
              {t('marginSelectorForm.belowMarket')}
            </MarketButton>
          </ButtonGroup>
        </FormGroup>

        <h3>{t('marginSelectorForm.sellPrice')}</h3>
        <div className="border rounded p-3">
          1 TODO = 1,234.00 {currency}
        </div>
        <small>{t('marginSelectorForm.priceOrigin')}</small>

        <h3>{t('marginSelectorForm.ourFee')}</h3>
        <div className="border rounded p-3">
          TODO SNT
        </div>

        {!this.props.margin && this.props.margin !== 0 && <p className="text-muted mt-3">{t('marginSelectorForm.enterMargin')}</p>}
      </Form>
    );
  }
}

MarginSelectorForm.propTypes = {
  t: PropTypes.func,
  margin: PropTypes.number,
  marketType: PropTypes.number,
  currency: PropTypes.string,
  marginChange: PropTypes.func,
  marketTypeChange: PropTypes.func
};

export default withNamespaces()(MarginSelectorForm);
