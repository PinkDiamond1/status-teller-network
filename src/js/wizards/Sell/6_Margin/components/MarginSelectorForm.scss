@import "../../../../../css/variable-overrides";
$slider-height: 10px;

.rc-slider {
  .rc-slider-rail {
    height: $slider-height;
    background-color: $secondary;
  }

  .rc-slider-track {
    height: $slider-height;
    background-color: $primary;
  }

  .rc-slider-handle{
    height: 30px;
    width: 30px;
    margin-top: -11px;
    border-color: $primary;
  }
}
