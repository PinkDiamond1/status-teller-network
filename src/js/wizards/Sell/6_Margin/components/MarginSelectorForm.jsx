/* global web3 */
import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {FormGroup, InputGroup, InputGroupAddon, InputGroupText, Col, Row} from 'reactstrap';
import Input from 'react-validation/build/input';
import {withNamespaces} from 'react-i18next';
import Form from 'react-validation/build/form';
import {isNumber, required, lowerThan} from '../../../../validators';
import Slider from 'rc-slider/lib/Slider';
import 'rc-slider/assets/index.css';
import './MarginSelectorForm.scss';

const ABOVE = 0;
const BELOW = 1;

class MarginSelectorForm extends Component {
  onMarginChange = (value) => {
    if (value < 0) {
      if (this.props.marketType === ABOVE) {
        this.props.marketTypeChange(BELOW);
      }
      value *= -1;
    } else {
      if (this.props.marketType === BELOW) {
        this.props.marketTypeChange(ABOVE);
      }
    }
    this.props.marginChange(value);
  };

  render() {
    const {t, currency, margin, marketType, token, prices, fee} = this.props;

    const basePrice = prices[token.symbol][currency];
    const marginPrice = (margin || 0) / 100 * basePrice;
    const calcPrice = basePrice + (marketType === ABOVE ? marginPrice : -marginPrice);
    const adjustedMargin = (marketType === ABOVE) ? margin : margin * -1;

    return (
      <Form ref={c => {
        this.form = c;
      }}>
        <h2>{t('sellerMarginContainer.title')}</h2>
        <FormGroup className="mb-0">
          <Row>
            <Col xs={9}>
              <Slider className="mb-3 p-4" min={-100} max={100} defaultValue={0}
                      onChange={(value) => this.onMarginChange(value)} value={adjustedMargin}/>
            </Col>
            <Col>
              <InputGroup className="full-width-input">
                <Input type="number"
                       name="margin"
                       id="margin"
                       placeholder="0"
                       className="form-control prepend"
                       value={adjustedMargin}
                       onChange={(e) => this.onMarginChange(e.target.value)}
                       validations={[required, isNumber, lowerThan.bind(null, 100)]}/>
                <InputGroupAddon addonType="append"><InputGroupText>%</InputGroupText></InputGroupAddon>
              </InputGroup>
            </Col>
          </Row>
          <p className="text-muted">A negative margin means you sell assets for less than what they are worth</p>
        </FormGroup>

        <h3>{t('marginSelectorForm.sellPrice')}</h3>
        <div className="border rounded p-3">
          1 {token.symbol} = {calcPrice.toFixed(4)} {currency}
        </div>
        <small>{t('marginSelectorForm.priceOrigin')}</small>

        {(fee || '0') !== '0' && <Fragment>
          <h3 className="mt-4">{t('marginSelectorForm.ourFee')}</h3>
          <div className="border rounded p-3">
            {web3.utils.fromWei(fee, 'ether')} SNT
          </div>
        </Fragment>}

        {!margin && margin !== 0 && <p className="text-muted mt-3">{t('marginSelectorForm.enterMargin')}</p>}
      </Form>
    );
  }
}

MarginSelectorForm.propTypes = {
  t: PropTypes.func,
  token: PropTypes.object,
  prices: PropTypes.object,
  fee: PropTypes.string,
  margin: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  marketType: PropTypes.number,
  currency: PropTypes.string,
  marginChange: PropTypes.func,
  marketTypeChange: PropTypes.func
};

export default withNamespaces()(MarginSelectorForm);
