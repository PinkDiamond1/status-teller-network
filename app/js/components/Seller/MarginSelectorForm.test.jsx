import React from "react";
import { shallow } from "enzyme";

import MarginSelectorForm from "./MarginSelectorForm";

describe('MarginSelectorForm', () => {
  it('should render correctly', () => {
    const component = shallow(<MarginSelectorForm margin={{}} fiat={{}} onMarginChange={() => undefined}/>);
  
    expect(component).toMatchSnapshot();
  });
});
