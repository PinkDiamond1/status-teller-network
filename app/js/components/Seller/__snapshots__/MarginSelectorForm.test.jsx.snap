// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarginSelectorForm should render correctly 1`] = `
<Form(t)>
  <h2 />
  <FormGroup
    tag="div"
  >
    <Label
      for="margin"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    />
    <InputGroup
      className="full-width-input"
      tag="div"
    >
      <Control(s)
        className="form-control prepend"
        id="margin"
        name="margin"
        onChange={[Function]}
        placeholder="0"
        type="number"
        validations={
          Array [
            [Function],
            [Function],
          ]
        }
      />
      <InputGroupAddon
        addonType="append"
        tag="div"
      >
        <InputGroupText
          tag="span"
        >
          %
        </InputGroupText>
      </InputGroupAddon>
    </InputGroup>
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <ButtonGroup
      className="asset-btns market-btns"
      role="group"
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      />
      <Button
        active={true}
        color="secondary"
        onClick={[Function]}
        tag="button"
      />
    </ButtonGroup>
  </FormGroup>
  <h3 />
  <div>
    1 TODO = 1,234.00 
  </div>
  <small />
  <h3 />
  <div>
    TODO SNT
  </div>
  <p
    className="text-info"
  />
</Form(t)>
`;
